### ISSUE TITLE
Training nerfacto with nerfbuster data

### ISSUE THREAD

**Issue by lilygoli** [ethanweber/nerfbusters#5]
Hi,

Thanks for the amazing work. I ran into a problem trying to replicate baseline results. The commands provided for training the nerfacto models for baseline, has `--eval-mode` argument that doesn't seem to be supported in nerfstudio. When I run the command I get `ns-train: error: unrecognized arguments: --eval-mode`.
Is there a workaround for this without needing to restructure the whole dataset files? And what nerfstudio version was used that worked with this command?

**Comment by lilygoli** [comment]
I found out that the `nerfbusters-changes` branch of nerfstudio has a slightly different dataparser. I adopted that parser and my problem is solved.
However, I cannot reproduce the very noisy results of the baseline that is presented in the nerfbusters website. The noisy parts in my trained nerf appear to be much more cloudy and less dense. Could you please share if you used any other specific hyper-parameters like distortion loss weight that is different from nerfstudio default? Are there any other changes in `nerfbusters-changes` branch that might affect the baseline training quality? @ethanweber 
[Link](https://drive.google.com/file/d/1YXZmOyLGrhWmuOhQ-NmI-9g4xXXVbfSe/view?usp=sharing) to my rendered video of plant scene.
Thanks a lot!

**Comment by ethanweber** [comment]
Hi @lilygoli, thanks for reaching out. I think the artifacts actually look very similar, shown at the bottom of [this page](https://ethanweber.me/nerfbusters/supplement.html) for the plant scene. However, one difference is that when the distance is far away then we set the background to be white, as done [here](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/scripts/launch_nerf.py#L158). It should just be the Nerfacto default. However, we turn off camera optimization to make evaluation consistent.

You can take a look at our base config [here](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/nerfbusters/nerf/nerfbusters_config.py#L18) and the specific settings to turn off defaults to get it back to Nerfacto [here](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/nerfbusters/nerf/experiment_configs/experiments_baseline.py#L62). You'll notice the camera optimization is off, unlike specified [here](https://github.com/nerfstudio-project/nerfstudio/blob/b03e9beccdee2807bea71783f42b71fad5f2992e/nerfstudio/configs/method_configs.py#L94) in the `nerfbusters-changes` branch.

**Comment by lilygoli** [comment]
Thanks for the reply! @ethanweber  
Why are you setting the far plane to 2 when using scene contraction? 
Are the reported metrics also evaluated on images with white background or is this just for the purpose of rendering? Also is this why the coverage metric for Nerfacto is reported to be <100% for most scenes? I am a bit confused about what exactly is the formula for the coverage metric, given that the baseline nerfacto does not use visibility mask.  In my setting I am defining coverage per image as the number of pixels with accumulation value (opacity) less than 0.5 divided by the total number of pixels in an image.

**Comment by ethanweber** [comment]
Hi! So the far plane of 2 will include all of the training cameras, since the [-1,1] box is defined to contain all the train and eval cameras. We set it to 2 because we only want to evaluate the foreground, and when things are deleted by various baselines, we don't want to show and evaluate on the random background colors.

The reported metrics are only evaluated on the masked regions, being within a distance of 2 wrt. the far plane but also the visibility mask. Here you can see [the code](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/scripts/launch_nerf.py#L194-L197). Here you'll notice that the visibility mask is actually the same for every baselines. We train a Nerfacto model on the train+eval images, and then we compute the visibility from the eval images using the median depth. It checks to see if the point is scene by training images or not. If it is, we want to evaluate all the methods there.

**Comment by lilygoli** [comment]
I see. Are the visibility masks in the dataset files? and when Nerfacto coverage is reported to be 0.89 that means 89% of the pixels in the visibility mask had depth less than 2?

**Comment by ethanweber** [comment]
~~It means that the number of visible pixels and the number of pixels with depth < 2 is 89% of the total pixels.~~ Edit: you are correct!

The masks are not in the dataset files. We train a "pseudo ground truth" model that uses both train and eval images. It's described in more detail in the paper. Essentially we created an evaluation procedure where the eval views are far from the train views, and we think it's only fair to evaluate on visible regions. That's the reason for this "pseudo ground truth" Nerfacto model.

**Comment by lilygoli** [comment]
oh I thought [this line](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/scripts/launch_nerf.py#L300) meant the denominator is the number of visible pixels. Okay that makes sense thank you!

**Comment by ethanweber** [comment]
Oh yes, yes you are right! It's the percent of visible pixels with depth < 2. Depth < 2 is the proxy for something being there or not. The denominator is always the same since the visibility mask is always the same.

**Comment by lilygoli** [comment]
I see. Is there any code on how you generate the visibility masks or any chance you could also upload that with the dataset?

**Comment by ethanweber** [comment]
So visibility code touches a few files on [the nerfbusters branch](https://github.com/nerfstudio-project/nerfstudio/pull/1786/files). The steps are there if you want to reproduce it. It adds a new field to the Nerfacto model. The field is initialized with cameras from the DataManager--with code in the custom pipeline file--which is why I didn't merge it into the main branch of Nerfacto. It would be a bit one-off as is. I'll look into providing the visibility masks for you.

**Comment by ethanweber** [comment]
Hey @lilygoli, I uploaded the visiblity masks (`visibility_median_count` from the pseudo ground-truth model) [here](https://drive.google.com/file/d/1Wy77GlKCF4V7Z4wb0lSfI7iFBTNOqHfu/view?usp=sharing). Just be aware that they are downscaled from the original images by a factor of 2, as per [this line](https://github.com/ethanweber/nerfbusters/blob/1f4240344ecff1313f6dfa7be5e06fe7d3e29154/scripts/launch_nerf.py#L160). I'd still recommend incorporating the visibility loss yourself for experiments, but hopefully this is helpful. Each pixel value corresponds to the number of times that the pixel's point in 3D (after backprojecting with median depth) is in the forward-facing frustum view of a training camera.

**Comment by ethanweber** [comment]
I'll close this now, but just get a hold of me if you need anything!

**Comment by lilygoli** [comment]
Hi again!

So I ended up reproducing the masks and I noticed (small) differences in the metrics from my runs and the paper results. I was wondering if the results in the paper are averaged over a number of runs? and if so how big is the variance? Just checking if I am getting correct results and this is just due to running on a different machine etc. . 

for example here are the results I get for the 'art' scene:

Processing experiment: art---nerfacto-visibility-cube ...

coverage: 0.7181332997103643
depth: 974.256577706144
disparity: 0.15155763385099405
lpips: 0.3608271160771727
normals: 59.56734950316019
normals_11.25: 0.04745309685445365
normals_22.5: 0.15361157821376897
normals_30: 0.23734036396002636
normals_median: 53.754623519641726
psnr: 14.346749822520676
ssim: 0.4153017530727653

Processing experiment: art---nerfacto---pseudo-gt ...

coverage: 0.8748574999457631
depth: 0.0
disparity: 0.0
lpips: 0.10465816219009501
normals: 0.007647333154194681
normals_11.25: 0.9999999900103947
normals_22.5: 0.9999999900103947
normals_30: 0.9999999900103947
normals_median: 0.0
psnr: 24.383406516559962
ssim: 0.8166800973801639

Processing experiment: art---nerfacto ...

coverage: 0.8636427475087469
depth: 1637.1690739340634
disparity: 0.2164474535504533
lpips: 0.3989156989078948
normals: 62.239795088102035
normals_11.25: 0.04544459727681216
normals_22.5: 0.1471749302728216
normals_30: 0.22700989021602289
normals_median: 56.19305057632191
psnr: 14.229257455751217
ssim: 0.3766709031672451 

Thanks!
